runtime/marshaler.go

```diff
@@ -15,8 +15,8 @@ type Marshaler interface {
 	NewDecoder(r io.Reader) Decoder
 	// NewEncoder returns an Encoder which writes bytes sequence into "w".
 	NewEncoder(w io.Writer) Encoder
-	// ContentType returns the Content-Type which this marshaler is responsible for.
-	ContentType() string
+	// ContentType returns the response Content-Type for "v".
+	ContentType(v interface{}) string
```

Comment:
This is going to be a painful breaking change to the API. Is there ANY way to do this without passing in v?

Reply:
hmm well the problem is that the content type right now is not dynamic. I suppose we could add another function like `ContentTypeForMessage(m proto.Message) string` and make that work as a preferred content-type setter. i.e. if `ContentTypeForMessage(m proto.Message) string` returns a non-empty string, we use it. otherwise we use the existing `ContentType() string`. now this has a similar problem as it changes a public interface, which would result in compilation errors for implementors anyway, so I'm afraid I don't have a clever solution. this feels like the lesser of two evils to me?

Reply:
bumping this - any thoughts here?

Reply:
Unfortunately adding another method to the interface suffers from the same API breaking change problem so I would rather go with the simpler breaking change if it is possible. Just spitballing cause I don't think it's possible, there isn't a way to make a type union or something, is there?

Reply:
actually wow yea I think it is possible. so via a type union I think we can do something like this - https://play.golang.org/p/G4x7sEtEHbh - which in this case would mean we have something like this probably in this file:

```
// ContentType returns the Content-Type header value given a Marshaler or TypeAwareMarshaler and the interface to marshal v.
func contentType(m, v interface{}) string {
  switch t := m.(type) {
  case TypeAwareMarshaler:
    return t.ContentTypeForInterface(v)
  case Marshaler:
    return t.ContentType()
  }
  return ""
}
```

then we'd update all the `marshaler.ContentType()` calls in this package to `contentType(marshaler, v)`. what do you think of that? I think that is not technically API breaking, right?